File: solutions/go/src/playground/algorithm/kth_smallest_element_in_a_bst.go
package algorithm

func kthSmallest(root *TreeNode, k int) int {
	current := 0
	result, _ := kthSmallestR(root, &current, k)
	return result
}

func kthSmallestR(node *TreeNode, current *int, target int) (value int, found bool) {
	if node == nil {
		return 0, false
	}

	if left, ok := kthSmallestR(node.Left, current, target); ok {
		return left, true
	}

	if *current = *current + 1; *current == target {
		return node.Val, true
	}

	if right, ok := kthSmallestR(node.Right, current, target); ok {
		return right, true
	}

	return 0, false
}

File: solutions/java/src/main/java/playground/algorithm/KthSmallestElementInABst.java
package playground.algorithm;

import playground.algorithm.common.TreeNode;

import java.util.Objects;

public class KthSmallestElementInABst {

    private int k;

    public int kthSmallest(TreeNode root, int k) {
        this.k = k;
        return Objects.requireNonNull(kthSmallestNode(root)).val;
    }

    private TreeNode kthSmallestNode(TreeNode node) {
        if (node.left != null) {
            TreeNode result = kthSmallestNode(node.left);
            if (result != null) {
                return result;
            }
            k--;
        }

        if (k == 1) {
            return node;
        }

        if (node.right != null) {
            k--;
            TreeNode result = kthSmallestNode(node.right);
            if (result != null) {
                return result;
            }
        }
        return null;
    }
}